### ISSUE TITLE
Tutorial: Train an Adaptive CSI Activity Classifier

### ISSUE THREAD

**Issue by ruvnet** [ruvnet/wifi-densepose#143]
# 🎓 Tutorial: Train an Adaptive CSI Activity Classifier

## Introduction

WiFi-based activity classification using ESP32 Channel State Information (CSI) traditionally relies on hand-tuned thresholds — brittle rules that break when you move the sensor to a new room. **ADR-048** introduces an adaptive classifier that *learns* your environment's signal characteristics from labeled recordings, replacing guesswork with data-driven classification.

This tutorial walks you through the full workflow: recording labeled CSI sessions, training a Rust-native logistic regression model, and verifying that your WiFi sensing system accurately distinguishes between **absent**, **present (still)**, **present (moving)**, and **active** states.

---

## What You'll Learn

1. **Record labeled CSI sessions** — capture "ground truth" data for each activity state
2. **Train the adaptive classifier** — build a 15-feature logistic regression model in pure Rust
3. **Verify classification accuracy** — check per-class statistics and overall accuracy
4. **Understand the smoothing pipeline** — how three-stage signal processing stabilizes noisy CSI

## Prerequisites

- ESP32-S3 running CSI firmware (see [User Guide — Hardware Setup](docs/user-guide.md))
- Rust sensing server running: `cargo run -p wifi-densepose-sensing-server -- --http-port 3000 --source auto`
- Browser open to `http://localhost:3000`

## Step 1: Record Training Data

Record 60–120 seconds of each activity state. The filename prefix determines the label:

```bash
# Empty room — leave the room completely
curl -X POST http://localhost:3000/api/v1/recording/start \
  -H "Content-Type: application/json" \
  -d '{"id": "train_empty_room"}'
# Wait 60-120 seconds with nobody in the room
curl -X POST http://localhost:3000/api/v1/recording/stop

# Still presence — sit or stand motionless
curl -X POST http://localhost:3000/api/v1/recording/start \
  -H "Content-Type: application/json" \
  -d '{"id": "train_still_reading"}'
# Sit still for 60-120 seconds
curl -X POST http://localhost:3000/api/v1/recording/stop

# Walking — move around the room
curl -X POST http://localhost:3000/api/v1/recording/start \
  -H "Content-Type: application/json" \
  -d '{"id": "train_walking_around"}'
# Walk normally for 60-120 seconds
curl -X POST http://localhost:3000/api/v1/recording/stop

# Active movement — exercise, jumping, etc.
curl -X POST http://localhost:3000/api/v1/recording/start \
  -H "Content-Type: application/json" \
  -d '{"id": "train_active_exercise"}'
# Do jumping jacks, stretches, etc. for 60-120 seconds
curl -X POST http://localhost:3000/api/v1/recording/stop
```

## Step 2: Train the Model

```bash
curl -X POST http://localhost:3000/api/v1/adaptive/train
```

This extracts 15 features per frame, z-score normalizes, and trains a multiclass logistic regression with softmax output. Training 3,000+ frames completes in <1 second.

## Step 3: Check Accuracy

```bash
curl -s http://localhost:3000/api/v1/adaptive/status | python3 -m json.tool
```

Look for `accuracy` (target: >60% with good recordings) and per-class `count` to verify balanced training data.

## Step 4: Test Live Classification

Walk around, sit still, leave the room — watch the dashboard or Observatory (`/observatory.html`) update in real time. The adaptive model blends 70% learned confidence with 30% threshold-based confidence for smooth transitions.

### 15-Feature Vector
The classifier extracts 7 server-computed features (variance, motion band power, breathing band power, spectral power, dominant frequency, change points, RSSI) plus 8 subcarrier statistics (mean, std, skew, kurtosis, IQR, spectral entropy, max, range) from each CSI frame.

### Three-Stage Smoothing
1. **Adaptive baseline subtraction** — EMA α=0.003 tracks quiet-room noise floor
2. **EMA + median filtering** — blended motion score, trimmed-mean vital signs
3. **Hysteresis debounce** — 4 consecutive frames required before state transition

### Vital Signs Stabilization
- 21-frame sliding window → trimmed mean (middle 50%) → EMA α=0.02
- Dead-bands: HR ±2 BPM, BR ±0.5 BPM — prevents display creep
- Outlier rejection: HR >8 BPM/frame, BR >2 BPM/frame discarded

## References

- [ADR-048: Adaptive CSI Activity Classifier](docs/adr/ADR-048-adaptive-csi-classifier.md)
- [User Guide — Adaptive Classifier section](docs/user-guide.md)
- [Observatory Visualization](docs/user-guide.md#observatory-visualization)

---

🤖 Generated with [claude-flow](https://github.com/ruvnet/claude-flow)
